### PR TITLE
feat: update dependencies, install angular-material and convert the a…

### PR DESCRIPTION
**Pull request by jzolnowski**
- update dependencies
- install angular-material and convert the application to be similar to angularjs-angular-hybrid-angular-cli 
- improve webpack configuration
- migrate to eslint

NOTE: the [e2e PR](https://github.com/xlts-dev/angularjs-webpack-babel/pull/19) should be merged first

### CODE REVIEW

File: src/app/version-stamp.component.js
Comment (gkalpak) [inline src/app/version-stamp.component.js]: OOC, should we do `$window.angular` (for consistency)?

File: src/app/app-angularjs.component.js
Comment (gkalpak) [inline src/app/app-angularjs.component.js]: Could we omit `controller` altogether?

File: package.json
Comment (gkalpak) [inline package.json]: Since we now have `npm-check` as a direct dependency, could this be just `"update": "npm-check -u"`?

File: config/plugins.mjs
Comment (gkalpak) [inline config/plugins.mjs]: Let's order imports alphabetically (by import path):

```suggestion
import {CleanWebpackPlugin} from 'clean-webpack-plugin';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import MiniCssExtractPlugin from 'mini-css-extract-plugin';
import webpack from 'webpack';
import { paths, isProd } from './helpers.mjs';
```

File: config/plugins.mjs
Comment (gkalpak) [inline config/plugins.mjs]: Let's make this more future proof by removing the version:

```suggestion
    title: 'XLTS for AngularJS with Webpack',
```

File: config/common.mjs
Comment (gkalpak) [inline config/common.mjs]: Let's order imports alphabetically (by import path):

```suggestion
import { plugins } from './plugins.mjs';
import { rules } from './rules.mjs';
```

File: config/common.mjs
Comment (gkalpak) [inline config/common.mjs]: I think it would be better to have a more descriptive name (both for the exported const and for the file itself).
Maybe `commonConfig` or `commonWebpackConfig` instead of just `common`.

This would give more context on the file's/code's purpose, without having to find the usages.

File: package.json
Comment (gkalpak) [inline package.json:77]: Why are we introducing AngularJS Material?
I thought we decided that we didn't want to introduce any new dependencies (including AngularJS Material).

File: src/index.html
Comment (gkalpak) [inline src/index.html]: Let's remove version numbers to make this more future proof.

File: src/index.html
Comment (gkalpak) [inline src/index.html]: OOC, why introduce `ng-csp`? It feels a bit outside the scope of this repo.
(If we decided to go ahead with this change, we should manually import `angular-csp.css`.)

File: gulpfile.ts
Comment (gkalpak) [inline gulpfile.ts]: How is `gulp` invoked in this repo? Is it run as part of `npm build`/`npm start` or should it be run manually?

File: .eslintrc.json
Comment (gkalpak) [inline .eslintrc.json]: What is the `go` global?

File: src/app/templates/templates.js
Comment (gkalpak) [inline src/app/templates/templates.js]: Should this file be checked in?
If we run the `templates` gulp task as part of the build (which we should imo), then could we git-ignore this file?

File: src/app/templates/index.js
Comment (gkalpak) [inline src/app/templates/index.js]: The `templates.js` file does not have any exports. What are we importing here?

File: src/app/version-stamp.component.js
Comment (gkalpak) [inline src/app/version-stamp.component.js]: It would be better to style images via CSS, because the hot module replacement seems to result in having some generic `img` styles (from AngularJS Material? - not sure) to override the `height` attribute with a `height: auto` CSS rule (which in turn results in the AngularJS image getting pretty big).

**Review by gkalpak — CHANGES_REQUESTED** [review]
For future PRs:
This kind of PR is usually much easier to review if it is broken up into multiple commits with one type of change per commit (for example, one commit with all changes related to updating dependencies, one commit related to improving the webpack config, one related to improving the ESLint config, etc.).

File: src/app/version-stamp.component.js
Comment (jzolnowski) [reply src/app/version-stamp.component.js]: Changed

File: src/app/app-angularjs.component.js
Comment (jzolnowski) [reply src/app/app-angularjs.component.js]: Removed

File: package.json
Comment (jzolnowski) [reply package.json]: Changed

File: config/plugins.mjs
Comment (jzolnowski) [reply config/plugins.mjs]: Reordered

File: config/plugins.mjs
Comment (jzolnowski) [reply config/plugins.mjs]: Version removed

File: config/common.mjs
Comment (jzolnowski) [reply config/common.mjs]: Reordered

File: config/common.mjs
Comment (jzolnowski) [reply config/common.mjs]: Changed to `webapck-common-config` to be consistent with `webapck-dev` and `webpack-prod` configs

File: package.json
Comment (jzolnowski) [reply package.json:77]: I know that such a comment was added to the tracking document. However, the changes I made were done before this document was written. I wanted the repository consistent with` angularjs-angular-hybrid-angular-cli`, but it only applied to `AngularJS `without `TypeScript`. The added code improved the project configuration, for example, adding `parserOptions `in the `eslint `configuration.

File: src/index.html
Comment (jzolnowski) [reply src/index.html]: Version removed

File: gulpfile.ts
Comment (jzolnowski) [reply gulpfile.ts]: Add a new script `"build:templates": "gulp templates"`. Now when running `start` or `build` script, it lunch first` npm run build:templates`

Comment (jzolnowski) [reply gulpfile.ts]: image

File: .eslintrc.json
Comment (jzolnowski) [reply .eslintrc.json]: Removed

File: src/app/templates/templates.js
Comment (jzolnowski) [reply src/app/templates/templates.js]: Added to `.gitignore`

File: src/app/templates/index.js
Comment (jzolnowski) [reply src/app/templates/index.js]: Good catch! I removed the entire templates folder and import cached templates directly in `app-angularjs.module.js`

File: src/app/version-stamp.component.js
Comment (jzolnowski) [reply src/app/version-stamp.component.js]: Good to know:) thank for the info! Moved image styles to CSS

File: src/index.html
Comment (jzolnowski) [reply src/index.html]: It's a good practice to not use `eval()`, `Function(string)`, and similar functions to dynamically create and execute code from strings as well as inline resources, such as inline `<script>` and `<style>` elements.

It's true that to use the `ngHide` or `ngCloak `directive, you need to import` angular-csp.css` manually, but in this project, we don't use it, so I don't know if it makes sense to import this file out of the box.

**Comment by jzolnowski** [comment]
@gkalpak PR ready for another round of review

File: src/index.html
Comment (gkalpak) [reply src/index.html]: You are right that it is a good practice in general, but I feel it might be outside the scope of this repo (whose purpose is to demonstrate how to set app a basic app).

I don't feel strongly about it, so leaving it up to you.
I do think, though, that if we keep `ngCsp` we should include `angular-csp.css` for a couple of reasons:
- It is a tiny file, so does not tangibly affect the loading of the page.
- It is not only about `ngCloak` and `ngHide`/`ngShow`. It also has styles for some kinds of animations (`.ng-anchor`, `.ng-animate-shim`).
- Even if we don't use one of the affected features directly, it is possible that one of our dependencies does.  
  (For example, the developer should not care to know if AngularJS Material relies on animations that require `.ng-anchor`/`.ng-animate-shim`.)
- Even if we don't rely on any of the features right now, it is easy to start relying in the future (either directly or indirectly via a new dependency).

File: package.json
Comment (gkalpak) [reply package.json:77]: I do have a preference for keeping things simple with these demo apps (i.e. not include extra dependencies unless they are necessary or we want to demonstrate something), but I don't feel too strongly about it.

I'll defer to @Splaktar, because we had discussed that we don't want to introduce a new dependency.
If this is fine with him, it's fine with me.

**Review by gkalpak — CHANGES_REQUESTED** [review]
Just a couple of comments.
Otherwise lgtm 👍

Great stuff :100:

File: package.json
Comment (gkalpak) [inline package.json]: I just realized that this command will not pick up template changes while it is running. I suggest we use something like [npm-run-all][1] to run `gulp watch` and `webpack-dev-server` in parallel.

[1]: https://www.npmjs.com/package/npm-run-all

File: src/app/version-stamp.component.js
Comment (gkalpak) [inline src/app/version-stamp.component.js]: We are setting the height in CSS, so we don't need it here:

```suggestion
      <span class="version"><img src="{{ $ctrl.angularJsIcon }}" alt=""> AngularJS v{{ $ctrl.versions.angularjs.full }} ({{ $ctrl.versions.angularjs.codeName }})</span>
      <span class="version"><img src="{{ $ctrl.angularjsMaterialIcon }}" alt=""> AngularJS Material v{{ $ctrl.versions.md }}</span>
```

File: package.json
Comment (gkalpak) [inline package.json]: Nit: I would name this something more descriptive, such as `watch:templates`.

**Comment by jzolnowski** [comment]
@gkalpak PR ready for another round of review

File: package.json
Comment (gkalpak) [inline package.json]: Nit: This can be shortened as:

```suggestion
    "start": "npm run build:templates && run-p start:server gulp:watch",
```

**Review by gkalpak — CHANGES_REQUESTED** [review]
Almost ready :tada:

**Comment by jzolnowski** [comment]
@gkalpak PR ready for final review 😄

**Review by gkalpak — APPROVED** [review]
LGTM (pending a decision on [this discussion][1] - I've pinged Michael to take a look at it) 🚀

[1]: https://github.com/xlts-dev/angularjs-webpack-babel/pull/18/files#r1124280976

File: package.json
Comment (Splaktar) [reply package.json:77]: I agree with @gkalpak, and I specifically spoke with @jzolnowski and let him know that we should not add AngularJS Material to this repo.

That said since the work is already done, I will look at it and decide if we should keep it. I will need to run the app locally to do that, so I won't be able to take care of that right now.

**Review by Splaktar — CHANGES_REQUESTED** [review]
- [x] I get the following errors when running `gulp templates`:

```
> gulp templates

[22:31:11] Failed to load external module ts-node/register
[22:31:11] Error: Cannot find module 'ts-node/register' from '/Users/splaktar/Git/xlts/angularjs-webpack-babel'
[22:31:11] Failed to load external module typescript-node/register
[22:31:11] Error: Cannot find module 'typescript-node/register' from '/Users/splaktar/Git/xlts/angularjs-webpack-babel'
[22:31:11] Failed to load external module typescript-register
[22:31:11] Error: Cannot find module 'typescript-register' from '/Users/splaktar/Git/xlts/angularjs-webpack-babel'
[22:31:11] Failed to load external module typescript-require
[22:31:11] Error: Cannot find module 'typescript-require' from '/Users/splaktar/Git/xlts/angularjs-webpack-babel'
[22:31:11] Failed to load external module sucrase/register/ts
[22:31:11] Error: Cannot find module 'sucrase/register/ts' from '/Users/splaktar/Git/xlts/angularjs-webpack-babel'
[22:31:11] Failed to load external module @babel/register
[22:31:11] Error: Cannot find module '@babel/register' from '/Users/splaktar/Git/xlts/angularjs-webpack-babel'
[22:31:12] Using gulpfile ~/Git/xlts/angularjs-webpack-babel/gulpfile.ts
[22:31:12] Starting 'templates'...
[22:31:12] Finished 'templates' after 17 ms
```
Is this expected?

- [x] The footer text and images aren't properly aligned. This isn't a problem in the angularjs-angular-material-hybrid demo.

Screenshot 2023-03-09 at 22 36 52

File: package.json
Comment (Splaktar) [reply package.json:77]: Based on the lack of functionality in the original version of this app and the fact that this work is already done, I think it is acceptable to add this additional dependency.

**Comment by jzolnowski** [comment]
Fixed `gulp templates` errors and fixed styles in the `versionStampComponent`

**Review by Splaktar — APPROVED** [review]

File: src/index.html
Comment (Splaktar) [reply src/index.html]: This still says "Webpack 5".

File: src/styles.scss
Comment (Splaktar) [inline src/styles.scss]: Why isn't this combined?

**Review by Splaktar — CHANGES_REQUESTED** [review]

**Comment by Splaktar** [comment]
Please rebase and resolve conflicts.

**Comment by jzolnowski** [comment]
Branch up to date with master

File: src/index.html
Comment (Splaktar) [reply src/index.html]: It looks like `ng-csp` was removed. Resolving this.

**Review by Splaktar — DISMISSED** [review]
- [x] Please respond to https://github.com/xlts-dev/angularjs-webpack-babel/pull/18#discussion_r1135800244. I know it's auto hidden by GitHub 😦

**Comment by jzolnowski** [comment]
> * [ ]  Please respond to [#18 (comment)](https://github.com/xlts-dev/angularjs-webpack-babel/pull/18#discussion_r1135800244). I know it's auto hidden by GitHub 😦

Sorry, I did not notice that the version repeats in several places - previously I removed it from the title. Now I have removed from the description and keywords

**Comment by jzolnowski** [comment]
I also set the `engine-strict` flag in `.npmrc`. I updated the npm and node versions in the `package.json` engines: `npm v9`, `node>=18`. I tested if the package installation, start and testing of the application works correctly. Everything looks fine

File: config/plugins.mjs
Comment (gkalpak) [inline config/plugins.mjs]: AFAICT, [since `webpack-dev-server` v4.0.0][1] (where `devServer.hot` defaults to `true`), we no longer need to include this plugin. (It is automatically included.)
Leaving it produces a warning in the terminal.

[1]: https://webpack.js.org/guides/hot-module-replacement/#enabling-hmr

File: config/webpack.test.js
Comment (gkalpak) [inline config/webpack.test.js:1]: I would add a comment here explaining why we need this `.js` file (and can't reuse other configs).

File: config/webpack.test.js
Comment (gkalpak) [inline config/webpack.test.js]: This produces the following warning in the terminal:

```
karma-webpack does not currently support custom entries, if this is something you need,
consider opening an issue.
ignoring attempt to set the entry option...
```

I would remove it or comment it out and add a comment explaining why there is no `entry`.

File: karma.config.js
Comment (gkalpak) [inline karma.config.js]: This line shouldn't be needed. The `*.test.js` files are processed as part of `index.tests.js` afaict.

File: .npmrc
Comment (gkalpak) [inline .npmrc]: This is a bit to strict imo.
To my experience, leaving this off is a better DX.

File: src/app/tabs.component.test.js
Comment (gkalpak) [inline src/app/tabs.component.test.js]: We should not set the 3rd argument. It is used to simulate `bindToController` bindings, which there are none in this case.

This is covering up the fact that the component's controller definition is incorrect and `$digestCount` is not properly initialized in the controller as we would expect :grin:

File: karma.config.js
Comment (gkalpak) [inline karma.config.js]: Nit:
It is more idiomatic to leave `autoWatch: true` and `singleRun: false` by default (or just remove them altogether, since the default is to run in watch mode), because most of the time people will want to run the unit tests in watch mode.

Then, if you want to run the tests once and exit (as we do in the `test` script), you can pass the `--single-run` option.

File: src/app/tabs.component.js
Comment (gkalpak) [inline src/app/tabs.component.js]: This is not right. Classes are not hoisted in the same way that functions do. As a result, `controller: TabsCtrl` is the same as `controller: undefined`.

You should either move the class definition above the `tabsComponent` object initialization or define the class inline (as you do with `VersionStampCtrl`):

```suggestion
  controller: class TabsCtrl {
    ...
  },
};
```

File: src/app/tabs.component.test.js
Comment (gkalpak) [inline src/app/tabs.component.test.js]: This is also not right. I don't understand the purpose of it, but it does not seem to belong in this test :grin:

File: src/app/tabs.component.test.js
Comment (gkalpak) [inline src/app/tabs.component.test.js:22]: All in all, I would simplify this whole test file as follows:

```js
describe('TabsCtrl', () => {
  let $componentController;
  let $rootScope;

  beforeEach(angular.mock.module('AngularJSApp'));
  beforeEach(angular.mock.inject((_$componentController_, _$rootScope_) => {
    $componentController = _$componentController_;
    $rootScope = _$rootScope_;
  }));

  it('should be defined', () => {
    const ctrl = $componentController('appTabs');
    expect(ctrl.$digestCount).toBe(0);

    $rootScope.$digest();
    // Expect `2`, because AngularJS always runs on more digest at the end to
    // to make sure there are no new changes.
    expect(ctrl.$digestCount).toBe(2);
  });
});
```

**Review by gkalpak — CHANGES_REQUESTED** [review]

**Comment by jzolnowski** [comment]
@gkalpak Thank you for the great explanation of the causes of errors in `tabs.component.test`. 
Ready for another round 😄

File: karma.config.js
Comment (gkalpak) [inline karma.config.js]: This line shouldn't be needed. The `*.test.js` files are processed as part of `index.tests.js` afaict.

File: src/app/tabs.component.test.js
Comment (gkalpak) [inline src/app/tabs.component.test.js]:
```suggestion
    // Expect `2`, because AngularJS always runs one more digest at the end to
```

My bad. I had the same typo in [my comment][1] 😇

[1]: https://github.com/xlts-dev/angularjs-webpack-babel/pull/18#discussion_r1138760713

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: karma.config.js
Comment (gkalpak) [inline karma.config.js:14]: Don't we need to preprocess `src/main.js`?
I would expect `preprocessors` to be:

```
preprocessors: {
  'src/main.js': ['webpack'],
  'src/index.tests.js': ['webpack'],
}
```

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: karma.config.js
Comment (gkalpak) [reply karma.config.js]: To clarify, I am referring to line 17 (`src/**/*.test.js`), not line 15 (`src/main.js`) :grin:

File: package.json
Comment (gkalpak) [inline package.json]: The karma jobs require the templates to have been built. Can you make sure that (a) the templates are built before the jobs run and (b) the templates are watched and rebuilt when Karma runs in watch mode? :pray:

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: package.json
Comment (gkalpak) [inline package.json]: This doesn't look right. We should remove this property altogether, since there is no main script for this repo.

File: package.json
Comment (gkalpak) [inline package.json]: Why do we need to run both `lint-staged` and `npm run lint`?

File: package.json
Comment (gkalpak) [inline package.json]:
```suggestion
    "test:unit:single": "run-s build:templates \"test:unit -- --single-run\"",
```

File: package.json
Comment (gkalpak) [inline package.json]: Why not:
```suggestion
    "start:prod-local": "http-server dist -p 4200 -c-1 -b",
```

File: package.json
Comment (gkalpak) [inline package.json]: Now that the names of the scripts have changed, we also need to update `README.md` accordingly.

**Review by gkalpak — CHANGES_REQUESTED** [review]
@jzolnowski, please also make sure all commits are signed :pray:

**Comment by jzolnowski** [comment]
> @jzolnowski, please also make sure all commits are signed 🙏

All commits signed 🎉

**Comment by jzolnowski** [comment]
Branch rebased with `master`

File: package.json
Comment (gkalpak) [inline package.json]: Unless I am missing something, we shouldn't need explicitly building the templates here:

```suggestion
    "test": "run-s lint test:unit:single test:e2e",
```

File: package.json
Comment (gkalpak) [inline package.json:101]: These two should be placed above `eslint-plugin-import-order`.

File: README.md
Comment (gkalpak) [inline README.md]: To be a bit more clear, I might change this to something like:

```suggestion
For running the unit tests once
```

File: README.md
Comment (gkalpak) [inline README.md]: To be a bit more clear, I might change this to something like:

```suggestion
For running the unit tests in watch mode
```

File: README.md
Comment (gkalpak) [inline README.md]:
```suggestion
For linting
```

File: README.md
Comment (gkalpak) [inline README.md]:
```suggestion
For running all checks/tests for the app (lint, unit and e2e tests)
```

File: README.md
Comment (gkalpak) [inline README.md]: Not directly related to this PR, but shouldn't this be

```suggestion
npm run build
```

File: README.md
Comment (gkalpak) [inline README.md]:
```suggestion
npm run lint
```

File: README.md
Comment (gkalpak) [inline README.md]: This could also be:

```suggestion
npm test
```

**Review by gkalpak — CHANGES_REQUESTED** [review]

**Comment by jzolnowski** [comment]
@gkalpak thank you for the review. Ready for another round 😄

**Review by gkalpak — APPROVED** [review]
LGTM :tada:

File: src/styles.scss
Comment (gkalpak) [reply src/styles.scss]: Actually, this seems to still not be resolved.

**Review by gkalpak — CHANGES_REQUESTED** [review]

**Comment by jzolnowski** [comment]
`engines.node` field in `package.json` set to v16

**Review by gkalpak — APPROVED** [review]

**Pull request merged**
